middle_end/flambda2.0/from_lambda/lambda_to_flambda_primitives.ml

```diff
@@ -52,38 +52,127 @@ let bint_shift bi prim arg1 arg2 =
     (Binary (Int_shift (C.standard_int_of_boxed_integer bi, prim),
              unbox_bint bi arg1, untag_int arg2))
 
-let string_or_bytes_access_validity_condition str kind index : H.expr_primitive =
+(* offset to substract to a string length depending
+   on the size of the read/write *)
+let length_offset_of_size size =
+  let offset =
+    match (size : Flambda_primitive.string_accessor_width) with
+    | Eight -> 0
+    | Sixteen -> 1
+    | Thirty_two -> 3
+    | Sixty_four -> 7
+  in
+  Immediate.int (Targetint.OCaml.of_int offset)
+
+(* equivalent to the one in `cmm_helpers.ml` *)
+let max_or_zero size_int x =
+  let register_bitsize_minus_one =
+    H.Simple (Simple.const (Naked_immediate (
+        Immediate.int (Targetint.OCaml.of_int (size_int * 8 - 1)))))
+  in
+  let sign =
+    H.Prim (Binary (Int_shift (Naked_nativeint, Asr),
+                  x, register_bitsize_minus_one)) in
+  let minus_one =
+    H.Simple (Simple.const (Naked_nativeint (Targetint.of_int (-1))))
+  in
+  let sign_negation =
+    H.Prim (Binary (Int_arith (Naked_nativeint, Xor),
+                  sign, minus_one)) in
+  let ret =
+    H.Prim (Binary (Int_arith (Naked_nativeint, And),
+                  sign_negation, x)) in
```

Comment:
Please put "`in`" on a separate line, unless the `let` fits all on one line.  Also the indentation seems strange here -- I suggest only two spaces before "sign_negation" (so the "s" is under the "P" of "Prim").